src/EventLoop.php

```diff
@@ -0,0 +1,98 @@
+<?php
+namespace Interop\Async\EventLoop;
+
+/**
+ * A façade class that provides a common API for accessing the process global
+ * event loop implementation.
+ *
+ * This class is useful because it allows asynchronous packages to access a
+ * global `EventLoopInterface` instance without relying on implementation-
+ * -specific methods for doing so.
+ *
+ * @todo How many static facade methods should be provided? A few more useful
+ * ones, all methods part of `EventLoopInterface`, or maybe none?
+ */
+final class EventLoop
+{
+    /**
+     * @var EventLoopInterface The global event loop instance.
+     */
+    private static $loop;
+
+    /**
+     * Gets the global event loop.
+     *
+     * @return EventLoopInterface The global event loop instance.
+     *
+     * @throws LoopNotSetError Thrown if the global event loop has not been set.
+     */
+    public static function get(): EventLoopInterface
+    {
+        if (self::$loop === null) {
+            throw new LoopNotSetError();
+        }
+
+        return self::$loop;
+    }
+
+    /**
+     * Sets the global event loop.
+     *
+     * This method may only be called once in the lifetime of a program. Calls
+     * to the event loop that occur before this method is called will result in
+     * an error thrown, so this method should be called at the beginning of the
+     * program.
+     *
+     * Library packages should not typically call this method. When writing a
+     * concrete program, it is up to the application developer to choose what
+     * event loop implementation to use, and all libraries will use the same
+     * implementation.
+     *
+     * @param EventLoopInterface $loop Event loop instance to use as the global loop.
+     *
+     * @throws LoopAlreadySetError Thrown if the global event loop has already been set.
+     */
+    public static function set(EventLoopInterface $loop)
+    {
+        if (self::$loop !== null) {
+            throw new LoopAlreadySetError();
```

Comment:
I don't imagine `with()` to be the default method for actually using the event loop. I think the typical usage that makes sense is to use `EventLoop::onReadable()`, etc., wherever you need it, and then the entry point of the program looks like this:

```php
EventLoop::set(new MyLoop());
// or alternatively for some libraries...
MyLoop\init(); // calls `set()` at some point

// whatever you want

// run the event loop
EventLoop::run();
```

For testing, you can end out the test with `EventLoop::drop()` to discard the loop set.

Now `with()` is a totally different usage, and I'm not sure we need both methods or which one is better. `with()` can be used by itself like this:

```php
EventLoop::with(new MyLoop(), function () {
    // everything needs to be inside this function to work
});
```